core/src/main/java/io/grpc/ClientCallImpl.java

```diff
@@ -43,25 +47,44 @@
 import io.grpc.transport.HttpUtil;
 
 import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
 
+import javax.annotation.concurrent.GuardedBy;
+
 /**
  * Implementation of {@link ClientCall}.
  */
 final class ClientCallImpl<ReqT, RespT> extends ClientCall<ReqT, RespT> {
+  /**
+   * Parameters from https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md
+   */
+  private static final double INITIAL_BACKOFF_MILLIS = 1 * 1000;
+  private static final double MULTIPLIER = 1.6;
+  private static final double MAX_BACKOFF_MILLIS = 2 * 60 * 1000;
+  private static final double MIN_CONNECT_TIMEOUT_MILLIS = 20 * 1000;
+  private static final double JITTER = .2;
+
   private final MethodDescriptor<ReqT, RespT> method;
   private final SerializingExecutor callExecutor;
   private final boolean unaryRequest;
   private final CallOptions callOptions;
+  private ClientCall.Listener<?> callListener;
+  @GuardedBy("this")
   private ClientStream stream;
   private volatile ScheduledFuture<?> deadlineCancellationFuture;
-  private boolean cancelCalled;
-  private boolean halfCloseCalled;
+  private volatile ScheduledFuture<?> retryFuture;
   private ClientTransportProvider clientTransportProvider;
   private String userAgent;
-  private ScheduledExecutorService deadlineCancellationExecutor;
+  private ScheduledExecutorService scheduledExecutor;
+  private boolean initialBackoffComplete;
+  private double previousDelayMillis = INITIAL_BACKOFF_MILLIS;
+  private State state = State.INIT;
```

Comment:
@carl-mastrangelo maybe just revert the state change and just focus this PR on reconnecting? I suspect it will be easier to get it through review.  You can raise a separate issue if you want to propose a way of cleaning up channel state.